core/src/main/java/jenkins/security/ResourceDomainConfiguration.java

```diff
@@ -0,0 +1,258 @@
+
+    private String url;
+
+    public ResourceDomainConfiguration() {
+        load();
+    }
+
+    @Override
+    public boolean configure(StaplerRequest req, JSONObject json) throws FormException {
+        req.bindJSON(this, json);
+        save();
+        return true;
+    }
+
+    @POST
+    public FormValidation doCheckUrl(@QueryParameter("url") String resourceRootUrlString) {
+        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+
+        return checkUrl(resourceRootUrlString, true);
+    }
+
+    private FormValidation checkUrl(String resourceRootUrlString, boolean allowOnlineIdentityCheck) {
+        if (ExtensionList.lookupSingleton(RootUrlNotSetMonitor.class).isActivated()) {
+            // This is needed to round-trip expired resource URLs through regular URLs to refresh them,
+            // so while it's not required in the strictest sense, it is required.
+            return FormValidation.warning(Messages.ResourceDomainConfiguration_NeedsRootURL());
+        }
+
+        resourceRootUrlString = Util.fixEmptyAndTrim(resourceRootUrlString);
+        if (resourceRootUrlString == null) {
+            return FormValidation.ok(Messages.ResourceDomainConfiguration_Empty());
+        }
+
+        if (!UrlHelper.isValidRootUrl(resourceRootUrlString)) {
+            return FormValidation.error(Messages.ResourceDomainConfiguration_Invalid());
+        }
+
+        if (!resourceRootUrlString.endsWith("/")) {
+            resourceRootUrlString += '/';
+        }
+
+        URL resourceRootUrl;
+        try {
+            resourceRootUrl = new URL(resourceRootUrlString);
+        } catch (MalformedURLException ex) {
+            return FormValidation.error(Messages.ResourceDomainConfiguration_Invalid());
+        }
+
+        String resourceRootUrlHost = resourceRootUrl.getHost();
+        try {
+            String jenkinsRootUrlHost = new URL(JenkinsLocationConfiguration.get().getUrl()).getHost();
+            if (jenkinsRootUrlHost.equals(resourceRootUrlHost)) { // TODO this only checks the host, do we care about port differences?
+                return FormValidation.error(Messages.ResourceDomainConfiguration_SameAsJenkinsRoot());
+            }
+        } catch (Exception ex) {
+            LOGGER.log(Level.CONFIG, "Failed to create URL from the existing Jenkins root URL", ex);
+            return FormValidation.error(Messages.ResourceDomainConfiguration_InvalidRootURL(ex.getMessage()));
+        }
+
+        StaplerRequest currentRequest = Stapler.getCurrentRequest();
+        if (currentRequest != null) {
+            String currentRequestHost = currentRequest.getHeader("Host");
+
+            if (currentRequestHost.equals(resourceRootUrlHost)) {
+                return FormValidation.error(Messages.ResourceDomainConfiguration_SameAsCurrent());
+            }
+        }
+
+        // TODO We could perform more elaborate permission checks to prevent users from setting a subdomain (not great wrt cookies?)
```

Comment:
> I really don't want to waste a bunch of time to investigate further to have a definitive answer

I think we need to provide a definitive answer. An administrative may not have easy access to a DNS entry that is not a subdomain.

> the cookies that matter are `HttpOnly`, so it shouldn't be a problem in practice

Additionally I see that `JSESSIONID` gets set with no `Domain` attribute, which according to [this](https://en.wikipedia.org/wiki/HTTP_cookie#Domain_and_path) (finding it hard to get the official reference here) means that using a subdomain should be safe.

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie seems to say the same thing. So as long as we're not using `Domain`, simple subdomains are fine.

To me this means that the "not the same host" requirement is sufficient. While different port/scheme might be a different _Origin_, for cookies it doesn't seem as clear cut (port isn't explicitly mentioned, and scheme would allow upgrading but not downgrading, but what if your reverse proxy is not configured correctly?) — In summary, we can't get wrong by requiring a different host, which should always be doable.

And finally, paths aren't a different _Origin_, so that's not good enough, even if a cookie `Path` is set.